tensorflow/compiler/xla/tests/llvm_compiler_test.cc

```diff
@@ -147,7 +155,11 @@ TEST_F(CpuCompilerTest, HooksTest) {
 }
 
 TEST_F(GpuCompilerTest, HooksTest) {
+#if TENSORFLOW_USE_ROCM
+  gpu::AMDGPUCompiler compiler;
+#elif GOOGLE_CUDA
   gpu::NVPTXCompiler compiler;
+#endif
   TestCompilerHooks(&compiler);
```

Comment:
@cheshire `gpu_compiler` is the base class. It is less meaningful for this test to test only the base class, but makes much more sense to test both child classes. On the other hand, if we are to test both child classes, I can't think of a way to make runtime switches, because there dependencies we cannot construct in the `NVPTXCompiler` side. If you insist this has to be avoided, I can make it a test on base class only. Let me know your preferences.

BTW, correct me if I'm wrong, I thought avoiding compile time switches is a rule in components but less so in tests.

Reply:
We don't really have a hard rule, since this is all fairly new.
My consideration is this: internally, we do *not* currently have a pre-submit which runs ROCm test. We have a pre-submit which runs a ROCm build, but it's not even triggered on every commit. So I can very well imagine a tiny well-intentioned change which accidentally breaks the ROCm build, since this is currently not triggered everywhere.
This creates a cascade of problems: your team will discover that the bot is broken, or the next person editing the ROCm-specific files will notice that. Thus it is much better to always build both branches, avoiding such issues in the future.

But in this case I understand your point about not being able to compile NVPTXCompiler without CUDA. However, do you think it would be possible to share the entire code block without duplication in gpu_compiler?

Reply:
I have pushed a new commit so that we create a local dummy `GpuCompilerTest` instead of macroed out of either `amdgpu` or `nvptx` compilers. This is the minimal change-list I can think of in order to make the test target independent. I'll let you judge which one you prefer so that we can either land in the first commit or the new commit. It depends on how much of the test coverage you would like this unit test to perform. 

- If you'd prefer the unit test to show-case about the minimal subset implementation of a gpu compilation backend, then the new commit justifies this need.
- If you'd prefer the unit test to actually regress on the actual implementation of both gpu compilers, the first commit make sure it is regressed on `amdgpu`/`nvptx` targets.

Reply:
Sorry, that wasn't my question.

My question is, why is the method `OptimizeHloPostLayoutAssignment` distinct for NVPTXCompiler and AMDGPUCompiler in the first place?

Reply:
Good point. Looks like the only differences between `NVPTXCompiler` and `AMDGPUCompiler` are two passes: `CublasGemmPadForTensorCores` and `GpuTreeReductionRewriter`. Looks like as long as I can make sure the first pass is not invoked at runtime at AMD side, the `OptimizeHloPostLayoutAssignment` routine can be merged. Will double check and let you know.

Reply:
Leaving an update here after I studied the newly added passes in `NVPTXCompiler` side. Turns out that there are three newly added passes after the last time we have it updated: `CublasGemmPadForTensorCores`, `GpuTreeReductionRewriter`, and `GemmAlgorithmPicker`. Two of which we can not support at the moment.

- `CublasGemmPadForTensorCores`: As long as I know, there's no gpu in amd side that support tensor cores. Though there is no compile time concerns from this pass.
- `GemmAlgorithmPicker`: This depends on the implementation of `GetBlasGemmAlgorithms()` from the stream executor side (which is unimplemented on `ROCm` side). Besides, compile time wise, it requires redzone allocator. To bring this pass up will require the new feature to be approved.

My suggestion toward this review is to keep the diff minimized, and introduce the three passes corrrectly instead: `reduction_*` to `AMDGPU`'s `OptimizeHloPostLayoutAssignment`. I will add a commit to finish this. We can deal with new pass bring up and refactoring (or possibly merging `OptimizeHloPostLayoutAssignment`) in follow-up PRs. 

FYI @whchung